### ISSUE TITLE
catalog_apply failing because of 'heterogeneous objects'

### ISSUE THREAD

**Issue by cedricr** [r-lidar/lidR#757]
Hello,  

on some of the french IGN tiles, `catalog_apply` is failing with the following error:
> Error in `st_geometry<-.sf`(`*tmp*`, value = res) : 
  is.character(value) || inherits(value, "sfc") is not TRUE
In addition: Warning message:
The list returned by 'catalog_apply' contains heterogeneous objects. Merging is impossible. A list has been returned. 

Here's a minimal reproduction:
```r
library(lidR)

download.file("https://storage.sbg.cloud.ovh.net/v1/AUTH_63234f509d6048bca3c9fd7928720ca1/ppk-lidar/IQ/LHD_FXX_0569_6275_PTS_C_LAMB93_IGN69.copc.laz", "tile1.copc.laz")
download.file("https://storage.sbg.cloud.ovh.net/v1/AUTH_63234f509d6048bca3c9fd7928720ca1/ppk-lidar/IQ/LHD_FXX_0569_6274_PTS_C_LAMB93_IGN69.copc.laz", "tile2.copc.laz")

ctg <- readLAScatalog(c("tile1.copc.laz", "tile2.copc.laz"), filter = "-keep_first")
cbounds <- catalog_boundaries(ctg)
```

- those two tiles are a bit strange, in as much as this corresponds to a classified (military base) perimeter, which is why part of the tiles are blank (but which is also why I need to find the boundaries!). I've got not idea at the moment if the problem is confined to those kind of tiles.
- if I remove the keep_first filter, it works properly, at least for that example.

Analysing a bit further, if I run concave_hull on both tiles separately, one returns a Polygon, and the other a MultiPolygon

```r
ctg1 <- readLAScatalog(c("tile1.copc.laz"), filter = "-keep_first")
ctg2 <- readLAScatalog(c("tile2.copc.laz"), filter = "-keep_first")
```

```
catalog_map(ctg1, st_concave_hull)
```
> Geometry set for 1 feature 
Geometry type: MULTIPOLYGON
Dimension:     XY
Bounding box:  xmin: 569000 ymin: 6274000 xmax: 570000 ymax: 6275000
Projected CRS: RGF93 v1 / Lambert-93
MULTIPOLYGON (((569000.3 6274580, 569000.4 6274...

```
catalog_map(ctg2, st_concave_hull)
```

> Geometry set for 1 feature 
Geometry type: POLYGON
Dimension:     XY
Bounding box:  xmin: 569819 ymin: 6273000 xmax: 570000 ymax: 6274000
Projected CRS: RGF93 v1 / Lambert-93
POLYGON ((570000 6273988, 569999.8 6273988, 569...

**Comment by Jean-Romain** [comment]
Thank you for the reproducible example. I'll investigate.

**Comment by cedricr** [comment]
FWIW: I’ve tried to run catalog_apply on the tiles of several cities, without the "keep_first" filter, but it nearly always fail with the "heterogeneous objects" error anyway. I can provide other reproductible examples if you feel like you need them.

**Comment by Jean-Romain** [comment]
Fixed. With the default parameters of `st_concave_hull` and the first return only you reached a limit case where the polygon produced was invalid on one point. `lidR` handles invalid cases with `sf::st_make_valid()` but in this specific case it produced 2 polygons in a `MULTIPOLYGON`.

I fixed it by casting `MULTIPOLYGON` into `POLYGON` + a warning because your are not supposed to get more than 1 polygons.

Now a better way to handle this case would be to change the parameters. Your polygons are overly complex and this brings nothing. Change the parameters to produce less detailed polygons e.g.:

```r
cbounds <- catalog_boundaries(ctg, concavity = 4, length_threshold = 10)
```

Problem gone and simpler/better/faster result
